### PR TITLE
IVostokHostExtensions.GetAll(bool, bool)

### PR DESCRIPTION
**Pull request by anaym**

### CODE REVIEW

File: Vostok.Hosting.Abstractions/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -179,4 +179,5 @@ Vostok.Hosting.Abstractions.Requirements.RequiresPort
 Vostok.Hosting.Abstractions.Requirements.RequiresPort.RequiresPort() -> void
 Vostok.Hosting.Abstractions.Requirements.RequiresSecretConfiguration
 Vostok.Hosting.Abstractions.Requirements.RequiresSecretConfiguration.RequiresSecretConfiguration(System.Type type, params string[] scope) -> void
-Vostok.Hosting.Abstractions.WellKnownApplicationIdentityProperties
\ No newline at end of file
+Vostok.Hosting.Abstractions.WellKnownApplicationIdentityProperties
```
Comment (kungasc) [inline Vostok.Hosting.Abstractions/PublicAPI.Shipped.txt]: New changes should be placed in Unshipped file, then `vostok-release` tool rotates it

File: Vostok.Hosting.Abstractions.Tests/RequirementChecker_Tests.cs [hunk 2]
```diff
@@ -127,6 +127,19 @@ public bool TryGet<TExtension>(string key, out TExtension result)
 
             public IEnumerable<(Type, object)> GetAll() 
                 => items.Select(pair => (pair.Key, pair.Value));
+
+            public IEnumerable<(Type Type, object Extension, string Key)> GetAll(bool withoutKeys, bool withKeys)
```
Comment (KungA) [inline Vostok.Hosting.Abstractions.Tests/RequirementChecker_Tests.cs:131]: What is this and why it in the tests?

File: Vostok.Hosting.Abstractions/IVostokHostExtensions.cs [hunk 3]
```diff
@@ -36,5 +36,10 @@ public interface IVostokHostExtensions
         /// Returns all types registered without key.
         /// </summary>
         IEnumerable<(Type, object)> GetAll();
+        
+        /// <summary>
+        /// Returns all types registered.
+        /// </summary>
+        IEnumerable<(Type Type, object Extension, string Key)> GetAll(bool withoutKeys, bool withKeys);
```
Comment (kungasc) [inline Vostok.Hosting.Abstractions/IVostokHostExtensions.cs]: You can't extend interfaces this way, it breaks backward compatibility

Comment (kungasc) [inline Vostok.Hosting.Abstractions/IVostokHostExtensions.cs]: Parameters are also unclear, we already have `GetAll()` which is equivalent to `GetAll(true, false)`

So we need only simple `GetAllKeyed()` method, which is already implemented [here](https://github.com/vostok/hosting/blob/6f4242854ae4ff7ba13dbedd150bd7341aa9e67e/Vostok.Hosting/Components/HostExtensions/HostExtensions.cs#L55).

**Review by kungasc — CHANGES_REQUESTED** [review]

File: Vostok.Hosting.Abstractions/IVostokHostExtensions.cs (hunk 3 above)
Comment (kungasc) [reply Vostok.Hosting.Abstractions/IVostokHostExtensions.cs]: Still actual

Comment (anaym) [reply Vostok.Hosting.Abstractions/IVostokHostExtensions.cs]: Where you found it?

File: Vostok.Hosting.Abstractions/IVostokHostExtensionsForKeyed.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+
+namespace Vostok.Hosting.Abstractions
+{
+    /// <inheritdoc cref="IVostokHostingEnvironment.HostExtensions"/>
+    [PublicAPI]
+    public interface IVostokHostExtensionsForKeyed : IVostokHostExtensions
```
Comment (kungasc) [inline Vostok.Hosting.Abstractions/IVostokHostExtensionsForKeyed.cs]:
```suggestion
    public interface IVostokHostKeyedExtensions : IVostokHostExtensions
```

File: Vostok.Hosting.Abstractions/IVostokHostExtensionsExtensions.cs [hunk 5]
```diff
@@ -13,5 +13,25 @@ public static class IVostokHostExtensionsExtensions
...
+        [NotNull]
+        public static IVostokHostExtensionsForKeyed AsForKeyed([NotNull] this IVostokHostExtensions extensions)
+            => extensions as IVostokHostExtensionsForKeyed ?? throw new InvalidOperationException($"Host extensions of type '{extensions.GetType().Name}' are not {nameof(IVostokHostExtensionsForKeyed)}.");
+
+        /// <summary>
+        /// <para>Attempts to convert given <paramref name="extensions"/> to <see cref="IVostokHostExtensionsForKeyed"/> for get all keyed.</para>
+        /// <para>Fails with <see cref="InvalidOperationException"/> if that's not possible.</para>
+        /// </summary>
+        [NotNull]
+        public static bool TryAsForKeyed([NotNull] this IVostokHostExtensions extensions, out IVostokHostExtensionsForKeyed forKeyed)
```
Comment (kungasc) [inline Vostok.Hosting.Abstractions/IVostokHostExtensionsExtensions.cs]: Seems useless, statement `extensions is IVostokHostExtensionsForKeyed keyed` is much more simple than call of this extension.

**Review by kungasc — APPROVED** [review]

**Pull request merged**